### PR TITLE
#22 - Adding package via HTTP

### PR DESCRIPTION
**Pull request by olegmoz**
Part of issue #22 
Adding package metadata by PUT method with JSON in body

### CODE REVIEW

**Comment by paulodamaso** [comment]
@g4s8 please assign me, I'm  available

**Comment by fanifieiev** [comment]
@0crat refuse

**Comment by 0crat** [comment]
This pull request [#40](https://github.com/artipie/composer-adapter/issues/40) is assigned to @fanifieiev[/z](https://www.0crat.com/u/fanifieiev), here is [why](https://www.0crat.com/footprint/CT2E6TK9B/8f3fada6-1534-4ba3-97e7-c20ad5fcf098); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @g4s8[/z](https://www.0crat.com/u/g4s8) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be a monetary reward for this job

**Comment by g4s8** [comment]
@0crat status

**Comment by 0crat** [comment]
> @0crat refuse ([here](https://github.com/artipie/composer-adapter/issues/40#issuecomment-598541748))

@fanifieiev The user @fanifieiev[/z](https://www.0crat.com/u/fanifieiev) resigned from [#40](https://github.com/artipie/composer-adapter/issues/40), please stop working. Reason for job resignation: Order was cancelled

**Comment by 0crat** [comment]
> @0crat refuse ([here](https://github.com/artipie/composer-adapter/issues/40#issuecomment-598541748))

@fanifieiev Job refused in 9 hours - no penalty, see [§6](http://www.zerocracy.com/policy.html#6)

**Comment by paulodamaso** [comment]
@g4s8 PLease assign me

**Comment by 0crat** [comment]
> @0crat status ([here](https://github.com/artipie/composer-adapter/issues/40#issuecomment-598590609))

@g4s8 This is what I know about this job in [CT2E6TK9B](https://www.0crat.com/p/CT2E6TK9B), as in [§32](http://www.zerocracy.com/policy.html#32):

  * The job [#40](https://github.com/artipie/composer-adapter/issues/40) is [in scope](http://www.zerocracy.com/policy.html#14) for 22hr
  * The role is `REV`
  * The job is not assigned to anyone
  * The [budget](http://www.zerocracy.com/policy.html#4) is not set yet
  * These users are banned and won't be assigned:
    * @olegmoz[/z](https://www.0crat.com/u/olegmoz): This user reported the ticket
    * @fanifieiev[/z](https://www.0crat.com/u/fanifieiev): User was resigned from the ticket
  * Job [footprint](https://www.0crat.com/footprint/CT2E6TK9B?q=%7Bjob%3A%27gh%3Aartipie%2Fcomposer-adapter%2340%27%2Ctype%3A%7B%24not%3A%2F%5ENotify%2F%7D%7D) (restricted area)

**Comment by 0crat** [comment]
This pull request [#40](https://github.com/artipie/composer-adapter/issues/40) is assigned to @dgarus[/z](https://www.0crat.com/u/dgarus), here is [why](https://www.0crat.com/footprint/CT2E6TK9B/2edacec0-410d-4e12-ae59-78cc9a8db6bc); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @g4s8[/z](https://www.0crat.com/u/g4s8) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be a monetary reward for this job

**Comment by dgarus** [comment]
@g4s8 there are minor comments

File: src/test/java/com/artipie/composer/http/PhpComposerTest.java [hunk 1]
```diff
@@ -131,12 +133,15 @@ void shouldFailPutPackageMetadata() {
...
-    @Disabled("Package upload is not implemented")
-    void shouldPutRoot() {
+    void shouldPutRoot() throws Exception {
         final Response response = this.php.response(
             "PUT /base",
             Collections.emptyList(),
-            Flowable.just(ByteBuffer.wrap("data2".getBytes()))
+            Flowable.just(
+                ByteBuffer.wrap(
+                    ByteStreams.toByteArray(new ResourceOf("minimal-package.json").stream())
```
[the remark this answers is not in the document]
Comment (olegmoz) [reply src/test/java/com/artipie/composer/http/PhpComposerTest.java]: @dgarus I think code is fine as it is, do not see any reason to use more classes from Cactoos

File: src/main/java/com/artipie/composer/http/Root.java [hunk 2]
```diff
@@ -23,23 +23,60 @@
...
 
 /**
  * Root resource. Used as endpoint to add a package.
  *
  * @since 0.1
  */
 public final class Root implements Resource {
+
+    /**
+     * Storage to read content from.
```
[the remark this answers is not in the document]
Comment (olegmoz) [reply src/main/java/com/artipie/composer/http/Root.java]: @dgarus Good point, thank you! Fixed it

**Comment by dgarus** [comment]
@g4s8 LGTM

**Pull request merged**